src/data/matrix/basic.lean

```diff
@@ -637,16 +639,35 @@ begin
   split_ifs with h; simp [h, matrix.one_apply_ne],
 end
 
+lemma algebra_map_eq_diagonal (r : R) :
+  algebra_map R (matrix n n α) r = diagonal (algebra_map R (n → α) r) :=
+matrix.ext $ λ i j, algebra_map_matrix_apply
+
 @[simp] lemma algebra_map_eq_smul (r : R) :
   algebra_map R (matrix n n R) r = r • (1 : matrix n n R) := rfl
 
+lemma algebra_map_eq_diagonal_ring_hom :
+  algebra_map R (matrix n n α) = (diagonal_ring_hom n α).comp (algebra_map R _) :=
+ring_hom.ext algebra_map_eq_diagonal
+
+@[simp] lemma map_algebra_map (r : R) (f : α → β) (hf : f 0 = 0)
+  (hf₂ : f (algebra_map R α r) = algebra_map R β r) :
+  (algebra_map R (matrix n n α) r).map f = algebra_map R (matrix n n β) r :=
+begin
+  rw [algebra_map_eq_diagonal, algebra_map_eq_diagonal, diagonal_map hf],
+  congr' 1,
+  dsimp,
+  simp_rw hf₂,
+  refl,
```

Comment:
No difficult rewrites needed if we use `ext`:
```suggestion
  ext m,
  simp only [hf₂, pi.algebra_map_apply]
```

Reply:
Or better:
```lean
  congr' 1 with x,
  simp only [hf₂, pi.algebra_map_apply]
```
since `congr'` has `ext` builtin.